### ISSUE TITLE
CSS parser error on invisible character <0x200b>

### ISSUE THREAD

**Issue by gabriel-vivas-sonarsource** [SonarSource/SonarJS#3667]
| Label | Description |
| --- | --- |
| Subject | CSS parser error on invisible character <0x200b> |
| Community | https://community.sonarsource.com/t/unexpected-unknown-type-selector/81316 |

### Explanation

User reported an unclear error message appearing on otherwise normal CSS code:

```css
.header-text-longname {
  display: none;
}​
.header-text-shortname {
  display: inline;
}​
```

> Unexpected unknown type selector "​"

This error is raised by Sonar when an invisible character `<0x200b>` is placed where a selector is expected. In this case, on lines 3 and 6 after the braces.

The `<0x200b>` (zero-width space, ZWSP) character is in the range `[0x2000 - 0x206F]` [General Punctuation](https://en.wikipedia.org/wiki/General_Punctuation).

Stylint also raises an error, as can be confirmed in the online playground, see working example [1].
The error message changes depending on the placement of the invisible character.

> Unable to lint CSS: SyntaxError: The string did not match the expected pattern.

> Unknown word (CssSyntaxError)

### Suggestion

(Perhaps also in JavaScript and TypeScript):

* Verify runtime behavior vs parser behavior for invisible characters to see if it could produce bugs.
* Consider transparently stripping/ignoring invisible characters to avoid parsing errors.
* Consider warning about the presence of an invisible character (instead of an unclear error).
* Consider implementing a new rule in SonarText that raises on invisible characters that may cause parsing errors. There is [precedent for this](https://rules.sonarsource.com/text/RSPEC-6389) and perhaps security implications.

[1] [Stylelint playground example](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0ALGBDWAnAtAFxgA88cAbCAOwHML0BbGAAmAB0LHGoBLAZwAdS6AJ7xGFSjADcbAL6BoAjZtUGbPiIkeKCFjy0GzNh279BIxlwqkLU2QoogANCB5CKedIQTgePR+EoAZlxUXqzsjCwgWACupDA8kaJhHByR6BTi7nhclDjiONEUANbiAO4UiYx4MTAOhilpJDFxeRAFxWUVCFU1deGpIABG5GBFrTgwdHx4QpXV0bX1A5DkuPkWAG7oVlA4aIRzvUsR-nQMbuOT07Pd84v9J2DRPHgQdDh8WBB8MDpC43ReDwLFQcFtcAFCmBspRDgs+ikTrAwIIsOgYRQcMMIKNxlBogIuGB0TAcE8Xm8Pl8fjouPE4fdEZFkaj0TlMdjcfl8YTiQQqd9ftl6d0ANrHDh3BGIgwPJkgYLiLAwSriuUyyKQCg8GBPbIbUk8qx8+I4UpcPAoHDcAIBX4wC5bUgLBIgCUpAC67pkxw90oGLPQaIxWJGY3ymm0lvSu0+gr+OAgBqwWC4sFdoilx0iAUoJAC9C4pH+3IJxpJeXoIszRweObzOALgOLAKBIJwVAdvyJjcLLaKMCEpW0UAZ-pOkIo0PZZO2YHGhX46DAMF28fR2jH2ZAk+nuSsFAwuCoaO4DpI+XE2vcFCgQd23GVe66Nfh293IfyhRKEHKW7rIADkIARogwoY4uGbRGkSFY6nE0LaBmPRvgBQEgVW1q6qyn5tFwUxRukeD-vKDDcOgjYYHg0TKpWYFfh0v4vshjINCAeirh2qa7EGGA8OMmzbGmxGsdy8Qrre7b8LqXBBGAFo3K+LEDKWvIVuiOB4XwUY4M01bMeOkQqeW-JwbqrxYEhWYAfkVwzDgPAQNRK7CcpbTqGi9mTESEDkNqLknOsFBOmm1qOcMpI8IImhkm85x4JZtbyoFwWxhAwIhupmnabp-mGbhKYwFQsRBmaKAWvEfDLiqtyJaxcY0nZ9E-n+NUofKpkIbgfA6tEUBtCi6A+AuDEtYpBnODA8HmR8PV9RMcRxcNzVMVZ7WTWZ2j4EIPxLZ0qruqtMrJDKAyKto1WiGqJ3yuSrzvOtcWuu6HBeuqPoPH624vKm1DjIepQHhd+nboUFq7YxuUgNsfAoORToLK0WTsvt6qRDS4nuJ2kTjhwx0nZERArtMAAK1JCnSSFXddAzfMu8nY89rFBKQpCJpVckzAz6rygE5DQGzdOc44jMDBSfACxzNw4xqzjVBAA4S-Tbrc4wr0ne9iKfQB6m6RMUx2YDWIwLmyoo-jUOkKUwiutLeMy4TMDTGb1MnJyRRxEN6ABAQuA8Bh7HgaMHs2yLE5cBZF7xAQo7K9TWvU5EZ2m2KodpN7vwxWc56RIzasyhrnrjTrsSkgHxI6pUkTkKUvxc-KxctAHS4rjgXs+5XFtW0IfHAtQLSA3XrEN6XbQGrem2fMbXAHK1SknMPnmAisuT-Ybbe190aSW9bg8DG7ZLkL3oKDGiLe2f8huDMb50d4eya767YYH2l7Yn1VeQwAD1it+nWAd9sXcQ4AX3iiF+v034t1Xt-K+JsgZb0ATgOgsRshkGsA-SIIDD6v1PhFdmpIYE303p3a29kQT9zQcLYBT9BQUGwe-KBh4f7tyIQAkhSDSAoIHpQ+U+8aF0Jbs3Uk68-4sO3t3UhfdSRcPGrwn4tDwE4Psngo2sD-5iKATdHym0PwzkyGyWERC6DQF+F0cay9cD7FnBXIh1cN5mK0RYogZAHRUEtB3SMOh0GnEWufVBjCCHJ0uu6eBO9uFHUZgTQgRMiIpxVgMIIEcP4vFXDnFW8drqJxoIQoJcSTgvCEHEA8JBIBnBjE9NJ3pfT2KzhcUoZUCCCI0tqNMcDiHdy8bdSk9UhT-F8Zfa+gTGBU3ruoh+uMIkgAdk7WJLt56-zJM8O6AoGpS1TjucOLwknR1SXHaWp0smDOGRkqG8ySlxTGTLCwwJYASKoOQ-xYYdnXTzoiAuL0zGLMpKRLg5FKp4B9kxI5AwAB6AAKUU5EABeHoIU4EhQABhwAATg9AAKgAJSgpwLChFyKPQAGp0WooACQXOAG81WHyKTvG6QmP5ALnbyjBTimFULEUooxVinF7KCVEtJWExE5KqnbkDMGGcgx0igkacI2+MB74CqRFhIM+iOSSqUe-AJrTWHtIVcyJVYrcj7x1EvLRmIGFCN-mohB7DOEUPGqKlVgcxjGu8r5dVLcZWiIQUfe51V7X6sdUaryy9MSNM1bK+V-qBoGo5E-H1Ujv50A8JhaNKqkIAEYo3YXFU-ao6AiySWDaaq1oSs3KpDPvfIyp8S3kImQSgVBYa3mWeTPSh0AwBpDCGj+X9GGetECE8RNquB+L9SKztM5u3SstV6kh8bR1eIdV2017r8EDNaXfOx47U0hj6dA9djKh6jIVeMlWkTomHvNsIlN2aDF7KZhsyOySY653vZkpUQMgUy2vWc7OJ7LnNJufOy+jzY7PMqR9cauYLhNiLCWDCABHaIEAGklvEXU+0OldSxQdLAGOUGoTLtKT2teM6B1tL4sOhduqdyEcnbFci07mHke1T3MhCbDxeJ0bkM5jHlHhqIZukRBGpwhiTYQPW1xR1IXhSJ58tFSTl1abY4T746O5EqsqNwaAdR8XNU065Wr1GIOQSO6RanRMzk0+eHTppGlXJaRGjOwG7UWfk9RVmSGUN6UHRopm6nMR1PKkxjeLHj3jRQAjWAJ4YAjyRgYljkivFZR0IjFVHcPPJdvOedL3QABM400KgQiutTqaWMQd3Rjlzs7Rg5mgtGVTEQFhxYDXJYf4bsvFFarHpjC9KTGXpOMyqFrK4U8s5ditleLCUkrJRS9JAw4jUEtHC34bR6IWkhuJyTBtrAZvGttw2S3XEoEqOmvLsntzfPInadE1FFMrpCyIsLgCvHXYondmiU7+MHsE3KrdAF3u3aojRMug1lO-gByRVcPyPsg9Lv7UeuGJ7KiCDPMaV2Yc3covdj4PFtPxDs8ohzsAnOqcB1juHuO0TUFJOuaaT20N+YGEDnHNEac1fp5tMNv2Xulsx2RHASHfgX14MUhjJHv79vnsZqj5mKeC+F1gUXmzeOrqYaFmX3r2PUYO5TpXKvxfEZ56ov7kbtw2X1v8BJmyuEY+sm5HypVgt4MhvkQEPh2y4cTAEeyjksCQM-nb4G1loh0CvrgOImB2yQrW0zrxFAw8R8QcmyeclgQJcYAAFnGon8PGd8h5oLb9WPXwEptTqmTBMSmO4qa8bSxqSPx5dVR9PSGus93+N50M4JFGTMcLM659Udt5RTJiTk2ZOZH1bJSWBk6C3LkHM-Wsn9OG3BPPnxBzW40OrTXRNULggxoj8ggQOeK6uScbv++TtaU1Nr79TEf-kXPfbKOl4Za-Xjd-3-+Y-4-dOaQNxX8NVu8P9zcAJv9cAH9D9-8hdkNUNZ0dUd9StpoBohp+ssBAVe9hs4VRtcUOVMVJsxtps+U5thUICUDNoSlBgLAgD1d38KMv9KDcBqDaCGcftTc+ckCvpmDMIeBxI7wLhWDaBppLxchBFIZICNJYx98BsZkZYcDoVuU8UJtlCUUZt+VbZ5tkC79I8xdM4k1Jc+0yMtd+cKDdDUFVcJcTdskApP8aMpDttO89JLtzCNo1gm8+oW8p50cQ9b93DEZytkYiESkBAiAmCLDuoYBep+ooo+Ia8bFIcb9WIpCoiYiD5Bo+JrMCddML9AMr9wD-Cys0i5oHpzwrEIca5kjRZeCSi3IFpyju09EKsiE+pohwoIiAiZgdoEjyM68HCtAdAm1Ywq9G8sMa1BCSB4Y21apRYD9fovMEDyM2iOiaNQZilwda8kivFpicBmsRwKitiqidjtgEYrArDiN9MGDWN+9bVOMaNdjzijdDDGdEC2NJFddtxHj9C1cbDBkwCodWJvjNknCrdpNKhXD5Rdj8gx4vCqQfDBt20UgR9WIk4YAAA1U4vSL9HhCAGeXOLfT0eod6GQEAGQIAA)

**Comment by gabriel-vivas-sonarsource** [comment]
It turns out that the parsing error also happens in the browser.

This can obscure the next CSS rule after the character, creating a bug.

See this example:

| Actual | Expected |
| --- | --- |
| ![actual](https://user-images.githubusercontent.com/103954349/213454615-66629097-3cc9-455d-b2fd-2a66802e5b92.png) | ![expected](https://user-images.githubusercontent.com/103954349/213454543-e2db847f-5b6b-451f-974c-27197c0d9b2f.png) |

Code to reproduce:
```html
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Invisible bug</title>
<style>
.first {
  background: blue;
  font-family: sans-serif;
  font-size: 10em;
}
.second {
  background: red;
}​
.third {
  background: green;
}
.fourth {
  color: white;
}
</style>
</head>
<body class="first second third fourth">
  There's an invisible bug in the code.
  The background will not be green, but red.
</body>
</html>
```



Note: Tested with Safari 16.2 (18614.3.7.1.5), Firefox 108.0.2 (64-bit), Chrome 109.0.5414.87 (arm64) and on macOS Ventura 13.1 (22C65).

**Comment by saberduck** [comment]
As a first step to tackle this problem we could strip these invisible control characters from the stream to avoid parsing issues and log a warning for the user.

**Comment by ilia-kebets-sonarsource** [comment]
This requires some discussions with other squads and requires specification.

**Comment by ilia-kebets-sonarsource** [comment]
raising issues could be done in another way, but our analysis should be robust!

**Comment by ericmorand-sonarsource** [comment]
Verified locally against SonarQube 10.4, using the latest plugin artifact.
